### PR TITLE
fix: dont apply host-config gating to stable behavior

### PR DESCRIPTION
**Pull request by weihanglo**
### What does this PR try to resolve?

Related: #17200 

rust-lang/cargo#17123 made host artifacts stop picking up
`target.cfg(…).runner` and `target.cfg(…).linker`
even without `-Zhost-config`.
That silently changed stable behavior that
`target.cfg` no longer applied to proc-macro
and build script compilation under `--target <host>`,
while `target.<triple>` config still did.

The old gate reused `host_artifact_uses_only_host_config`,
which is for the documented rustflags dual behavior.
It wasn't for runner and linker.

This PR narrows the gate to `target-applies-to-host = false`
and restore to pre-#17123 stable behavior,
while keeping `-Zhost-config` fix as it is still nightly.

### How to test and review this PR?

I've added a lot of tests because I don't want to miss any edge cases again.

Here is the change table.

| test | 1.97 | 1.98-beta (rust-lang/cargo#17123) | this PR (rust-lang/cargo#17198) |
|---|---|---|---|
| `custom_runner_cfg_proc_macro_test_with_target` | applies | not | _applies_ |
| `target_cfg_linker_build_script_with_target` | applies | not | _applies_ |
| `target_cfg_linker_proc_macro_with_target` | applies | not | _applies_ |
| `target_cfg_linker_proc_macro_test_with_target` | applies | not | _applies_ |
| `target_cfg_runner_build_script_with_host_config_and_target` | applies | not | not |
| `target_cfg_linker_build_script_with_host_config_and_target` | applies | not | not |

All other added tests behave identically across 1.97, 1.98-beta, and this PR.
See rust-lang/cargo#17200 for the runner/linker inconsistency they document.

### Note

We'll need to backport this to 1.98-beta once it lands.

### CODE REVIEW

**Comment by RalfJung** [comment]
This means we'll have to revert https://github.com/rust-lang/miri/pull/5121/changes/afd5aff4d1c588cfb6d52dbb6d42069532543db2 in Miri, so please let me know when this lands in the main repo. Maybe we can even include the revert of that commit in the PR that syncs this update to rust-lang/rust?

---

What is the intended behavior of `target.runner` wrt proc macro tests in various flag configurations after this PR?
- IMO it should *never* apply since proc macro tests clearly are a host artifact. Even cargo agrees with that: cargo passes `--target` to rustc for target artifacts, but does not pass `--target` to proc macro tests. So IMO with this PR cargo becomes internally inconsistent, where proc macro tests are the only kinds of crates that get built *without* `--target` but they get run *with* `target.runner`. That's why Miri needs special hacks to work around this inconsistency.
- I could live with it *always* applying `target.runner`. We'll just have to keep our special hack in Miri then.
- But if it gets sometimes applied and sometimes not (depending on `-Z` flags or so), then Miri is screwed. (Well maybe there's something complicated we can do depending on the exact circumstances of what this looks like. But it sure won't be pretty.)

File: tests/testsuite/tool_paths.rs [hunk 1]
```diff
@@ -442,6 +444,397 @@ Caused by:
...
+        "#,
+        ),
+    );
+
+    p.cargo("test --lib -v -Ztarget-applies-to-host -Zhost-config --target")
+        .arg(alternate)
+        .masquerade_as_nightly_cargo(&["target-applies-to-host", "host-config"])
+        .with_stderr_data(str![[r#"
+...
+[RUNNING] `[ROOT]/foo/target/debug/deps/foo-[HASH][EXE]`
```
Comment (RalfJung) [inline tests/testsuite/tool_paths.rs:774]: Am I misunderstanding this test, or does this not use any of the runners? That seems very strange?

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/cargo/17198/0a28f7930c7b559c37fc221347114f9c6434f2ae..4d02c28c2b39ee0a221b30b027a796e322c1e54c)*

Comment (weihanglo) [reply tests/testsuite/tool_paths.rs:774]: No, your are not. This is likely a bug.

> What is the intended behavior of target.runner wrt proc macro tests in various flag configurations after this PR?

Some purposes of this PR are

* to revert some stable behavior back to whatever 1.97 has, in order to give us a room to breath and discuss the intended behavior
* to capture the current state of the interaction around proc macro (tests)

See the counterpart PR against 1.97 that has the same issue <https://github.com/weihanglo/cargo/pull/92/changes>.

I am sorry about the whole mess. Will try to follow up with a new Cargo issue.

Comment (RalfJung) [reply tests/testsuite/tool_paths.rs:774]: > Some purposes of this PR are

That makes sense to me, thanks.

Comment (weihanglo) [reply tests/testsuite/tool_paths.rs:774]: Opened <https://github.com/rust-lang/cargo/issues/17200>.

---

> This means we'll have to revert https://github.com/rust-lang/miri/commit/afd5aff4d1c588cfb6d52dbb6d42069532543db2 in Miri, so please let me know when this lands in the main repo. Maybe we can even include the revert of that commit in the PR that syncs this update to rust-lang/rust?

Regarding this, I think we will backport this to 1.98-beta. Will kindly ask for your help to also revert that in beta when Cargo's beta backport PR merges. I am willing to do the Cargo submodule/ Miri subtree sync for both beta and nightly 🙇🏾‍♂️

Comment (RalfJung) [reply tests/testsuite/tool_paths.rs:774]: Miri is nightly-only so beta backports don't affect us.

**Review by epage — APPROVED** [review]
